### DISCUSSION TITLE
Cannot compile caribou successfully with SOFA

### DISCUSSION THREAD

**Discussion by nicklj** [jnbrunet/caribou#33] [Q&A] [answered]
As I posted in the SOFA forum(https://www.sofa-framework.org/community/forum/topic/question-on-tetrahedronhyperelasticityfemforcefield/), I still cannot compile caribou successfully with SOFA. The error message is as follows:

```
CMake Error at /home/liuj/opt/SOFA/master/lib/cmake/SofaGui/SofaGuiConfig.cmake:40 (find_package):
  Found package configuration file:
    /home/liuj/opt/SOFA/master/plugins/SofaGuiQt/lib/cmake/SofaGuiQt/SofaGuiQtConfig.cmake

  but it set SofaGuiQt_FOUND to FALSE so package "SofaGuiQt" is considered to
  be NOT FOUND.  Reason given by package:

  The following imported targets are referenced, but are missing: QGLViewer
```

Understand this maybe some problem related with SOFA. I can do some hack in the file SofaGuiQtConfig.cmake by setting 
`set(SOFAGUIQT_HAVE_QT5 1)`
while I'm not sure whether this will influence other places. 

After that, the cmake can finish, while when compiling, it complains:
```
/home/liuj/github/sofa/caribou/build/src/Topology/include/Caribou/Topology/IO/VTKReader.h:13:10: fatal error: vtkSmartPointer.h: No such file or directory
   13 | #include <vtkSmartPointer.h>

```
I assume this is because the make system cannot find the inlcude files for VTK. I'm using the vtk 7.1.1 install in ubuntu 20.04. So I further hack the CMakeLists.txt of caribou by adding:
include(${VTK_USE_FILE})

Now the compiling system complains:
```
/home/liuj/github/sofa/caribou/src/Plugin/Forcefield/TractionForce.cpp: In member function ‘virtual void SofaCaribou::forcefield::TractionForce::init()’:
/home/liuj/github/sofa/caribou/src/Plugin/Forcefield/TractionForce.cpp:72:56: error: ‘class sofa::component::topology::TriangleSetTopologyContainer’ has no member named ‘getTriangleDataArray’; did you mean ‘getTriangleArray’?
   72 |             d_triangles.setParent(&triangle_container->getTriangleDataArray());
      |                                                        ^~~~~~~~~~~~~~~~~~~~
      |                                                        getTriangleArray
/home/liuj/github/sofa/caribou/src/Plugin/Forcefield/TractionForce.cpp:79:48: error: ‘class sofa::component::topology::QuadSetTopologyContainer’ has no member named ‘getQuadDataArray’; did you mean ‘getQuadArray’?
   79 |             d_quads.setParent(&quad_container->getQuadDataArray());
      |                                                ^~~~~~~~~~~~~~~~
      |                                                getQuadArray

```
Appreciate that if anyone may help on this.

**Comment by jnbrunet** [comment, marked as answer]
Hey @nicklj 

Thanks a lot for bringing this discussion here, it might help others to fix their compilation issues in the future.

Can you tell me on which commit your caribou is on? You can go in the source code of caribou and copy and paste the output of `git log --oneline -n 1`.

Could you do the same for SOFA?

Finally, can you copy and paste the output of cmake when you compile the caribou plugin?

Normally, the master branch of Caribou should always compile with SOFA v20.12, and most often than not with the master branch of SOFA with maybe a delay of a couple of days to let me fix the code as SOFA get updated. So it is usually a good idea to stick to the v20.12 branch of SOFA unless you really need the latest updates. 

J-N

**Reply by nicklj** [reply]
Everything works fine on SOFA v20.12. Thanks a lot for the information.
Just a kind suggestion that you may want to mention what is the developing environment (i.e. SOFA version) for caribou in the README. This will provide valuable information and save a lot of time for some people like me who always want to trace the up-to-date development version of SOFA.
Anyway, thanks a lot for your work on this. Looking forward for more exciting features:)

**Reply by jnbrunet** [reply]
Awesome, and thanks a lot for the feedback, it is very appreciated. 

I'm actually planing to add results of nightly compilations for each versions of SOFA in the README. Hopefully this will avoid further issues like the one you just had.

Note that for the VTK issue, the fix should be merged soon.
